### DISCUSSION TITLE
automatically adjusting/zoomable map?

### DISCUSSION THREAD

**Discussion by merrygin** [mesa/mesa-geo#136] [Q&A]
I would like to display agents on a world map but only need a zoomed in part of it initially. Ideally, this map part would zoom out if necessary - ie, when an agent steps out of the bounds of the original map part (à la cartopy). 

As far as I understand, this is not possible with mesa out of the box, but does mesa-geo have a feature for that?

Thanks in advance for any help!

**Comment by wang-boyu** [comment]
You're right - in Mesa, a map is setup during initialization and remains fixed afterwards.

In Mesa-Geo, there's an auto zoom feature if no specific view & zoom level are specified: https://github.com/projectmesa/mesa-geo/issues/102. However, this is done only once. More specifically, it is controlled by the `hasFitBounds` variable in `MapModule.js`:

https://github.com/projectmesa/mesa-geo/blob/bea0f322e44a68e3f910ed9a4391aa7492812f68/mesa_geo/visualization/templates/js/MapModule.js#L35-L47

If it is not too much trouble, you may modify the source code and install your modified version, to remove this `hasFitBounds` variable and let the map do auto-zoom at each step, and see how it works. Just a reminder that you may also need to avoid any manual setting of the `view` and `zoom` when initializing the map.

@rht Does this sound good for a new feature? We can have another option for `MapModule`, so that it does auto-zoom at each step. By default it is turned off.

Edit: Or, we can check if the bounds have changed at each step to decide whether to auto-zoom or not. Then we don't need users to set any options.

**Reply by rht** [reply]
> Edit: Or, we can check if the bounds have changed at each step to decide whether to auto-zoom or not. Then we don't need users to set any options.

This sounds good to me as a default behavior, as long as the bounds checking doesn't get in the way of the rest of the code.

**Reply by wang-boyu** [reply]
Great. Let me turn this into an issue.

**Comment by merrygin** [comment]
Thanks a lot for these helpful suggestions and the quick reply! It may be too much trouble in this particular case, although for future extensions of our model we definitely should do that. I'll try it out, for sure, when I find the time!
Maybe just for a bit more context what our use case is: we want to map agents of an historical letter network, which starts quite regionally in Europe but gets more and more global after a while. So we don't need a whole world map for visualization initially, but later on we do.

**Reply by wang-boyu** [reply]
Sounds cool! All the best with your model : )
